Added an include file detailing `Dispose` and `DisposeAsync` lifetime bits (#23692)
dotnet/docs: docs/standard/garbage-collection/includes/disposables-and-dependency-injection.md

---

> author: IEvangelist

Reviewer: Most of our include files live in the includes directory at the root of the repo. Also, it's probably not a good idea to have relative links in an include file, since they depend on the including file being in a specific location.

Author: In the Azure docs, we started scoping include directories local to their relevant content area. This makes it a lot easier to manage. As for relative links, are you suggesting that we use absolute paths instead - or what alternative specifically?

Reviewer: I see there's an includes directory under docs/azure and docs/iot. If we put one here, I think it should go up a level to docs/standard. @BillWagner Are you okay with that?

> As for relative links, are you suggesting that we use absolute paths instead - or what alternative specifically?
Yes, use a path that starts with /dotnet/.

Author: I'm curious why you're want to move it to `docs/standard`? What is the benefit of that, I've always favored locally scoped includes - and I do not anticipate this include being needed elsewhere.

Reviewer: That was to match the level of the includes folders in the docs/azure and docs/iot directories.

Author: Ah, I see. I do not believe it is necessary with this case, unlike those directories - this include is really scoped to the articles at this level.

ms.topic: include
ms.date: 04/07/2021
ms.author: dapine
ms.custom: include
---

> [!TIP]

> > With regard to dependency injection, when registering services in an <xref:Microsoft.Extensions.DependencyInjection.IServiceCollection>, the [service lifetime](/dotnet/core/extensions/dependency-injection.md#service-lifetimes) is managed implicitly on your behalf. The <xref:System.IServiceProvider?displayProperty=nameWithType> and corresponding <xref:Microsoft.Extensions.Hosting.IHost?displayProperty=nameWithType> orchestrate resource cleanup. Specifically, implementations of <xref:System.IDisposable?displayProperty=nameWithType> and <xref:System.IAsyncDisposable?displayProperty=nameWithType> are properly disposed at the end of their specified lifetime.

Reviewer 2: I don't think you need "nameWithType" on these.

```suggestion
> With regard to dependency injection, when registering services in an <xref:Microsoft.Extensions.DependencyInjection.IServiceCollection>, the [service lifetime](/dotnet/core/extensions/dependency-injection.md#service-lifetimes) is managed implicitly on your behalf. The <xref:System.IServiceProvider> and corresponding <xref:Microsoft.Extensions.Hosting.IHost> orchestrate resource cleanup. Specifically, implementations of <xref:System.IDisposable> and <xref:System.IAsyncDisposable> are properly disposed at the end of their specified lifetime.
```

> > With regard to dependency injection, when registering services in an <xref:Microsoft.Extensions.DependencyInjection.IServiceCollection>, the [service lifetime](../../../core/extensions/dependency-injection.md#service-lifetimes) is managed implicitly on your behalf. The <xref:System.IServiceProvider?displayProperty=nameWithType> and corresponding <xref:Microsoft.Extensions.Hosting.IHost?displayProperty=nameWithType> orchestrate resource cleanup. Specifically, implementations of <xref:System.IDisposable?displayProperty=nameWithType> and <xref:System.IAsyncDisposable?displayProperty=nameWithType> are properly disposed at the end of their specified lifetime.

Author: ```suggestion
> With regard to dependency injection, when registering services in an <xref:Microsoft.Extensions.DependencyInjection.IServiceCollection>, the [service lifetime](/dotnet/core/extensions/dependency-injection.md#service-lifetimes) is managed implicitly on your behalf. The <xref:System.IServiceProvider?displayProperty=nameWithType> and corresponding <xref:Microsoft.Extensions.Hosting.IHost?displayProperty=nameWithType> orchestrate resource cleanup. Specifically, implementations of <xref:System.IDisposable?displayProperty=nameWithType> and <xref:System.IAsyncDisposable?displayProperty=nameWithType> are properly disposed at the end of their specified lifetime.
```

> > With regard to dependency injection, when registering services in an <xref:Microsoft.Extensions.DependencyInjection.IServiceCollection>, the [service lifetime](../../../core/extensions/dependency-injection.md#service-lifetimes) is managed implicitly on your behalf. The <xref:System.IServiceProvider?displayProperty=nameWithType> and corresponding <xref:Microsoft.Extensions.Hosting.IHost?displayProperty=nameWithType> orchestrate resource clean up. Specifically, implementations of <xref:System.IDisposable?displayProperty=nameWithType> and <xref:System.IAsyncDisposable?displayProperty=nameWithType> are properly disposed at the end of their specified lifetime.

Reviewer: ```suggestion
> With regard to dependency injection, when registering services in an <xref:Microsoft.Extensions.DependencyInjection.IServiceCollection>, the [service lifetime](../../../core/extensions/dependency-injection.md#service-lifetimes) is managed implicitly on your behalf. The <xref:System.IServiceProvider?displayProperty=nameWithType> and corresponding <xref:Microsoft.Extensions.Hosting.IHost?displayProperty=nameWithType> orchestrate resource cleanup. Specifically, implementations of <xref:System.IDisposable?displayProperty=nameWithType> and <xref:System.IAsyncDisposable?displayProperty=nameWithType> are properly disposed at the end of their specified lifetime.
```

>

> > For more information, see [Dependency injection in .NET](../../../core/extensions/dependency-injection.md).

Author: ```suggestion
> For more information, see [Dependency injection in .NET](/dotnet/core/extensions/dependency-injection.md).
```
